packages/flutter/test/material/text_field_test.dart

```diff
@@ -1764,6 +1764,164 @@ void main() {
+      await tester.pumpAndSettle();
+
+      sendKeyEventWithCode(20, true, true, false);         // DOWN_ARROW keydown
+      await tester.pumpAndSettle();
+
+      expect(controller.selection.extentOffset - controller.selection.baseOffset, 11);
+
+      sendKeyEventWithCode(20, false, true, false);          // DOWN_ARROW keyup
+      await tester.pumpAndSettle();
+      sendKeyEventWithCode(19, true, true, false);           // UP_ARROW keydown
+      await tester.pumpAndSettle();
+
+      expect(controller.selection.extentOffset - controller.selection.baseOffset, 0);
+    });
+
+
+    testWidgets('Down and up test 2', (WidgetTester tester) async{
+      await tester.pumpWidget(setupWidget());
+      const String testValue = 'a big house\njumped over a mouse\nOne more line yay'; // 11 \n 19
+      await tester.enterText(find.byType(TextField), testValue);
+
+      await tester.idle();
+      await tester.tap(find.byType(TextField));
+      await tester.pumpAndSettle();
+
+      for (int i = 0; i < 5; i += 1) {
+        sendKeyEventWithCode(22, true, false, false);             // RIGHT_ARROW keydown
+        await tester.pumpAndSettle();
+        sendKeyEventWithCode(22, false, false, false);            // RIGHT_ARROW keyup
+        await tester.pumpAndSettle();
+      }
+      sendKeyEventWithCode(20, true, true, false);               // DOWN_ARROW keydown
+      await tester.pumpAndSettle();
+      sendKeyEventWithCode(20, false, true, false);              // DOWN_ARROW keyup
+      await tester.pumpAndSettle();
+
+      expect(controller.selection.extentOffset - controller.selection.baseOffset, 12);
+
+      sendKeyEventWithCode(20, true, true, false);                 // DOWN_ARROW keydown
+      await tester.pumpAndSettle();
+      sendKeyEventWithCode(20, false, true, false);                // DOWN_ARROW keyup
+      await tester.pumpAndSettle();
+
+      expect(controller.selection.extentOffset - controller.selection.baseOffset, 32);
+
+      sendKeyEventWithCode(19, true, true, false);               // UP_ARROW keydown
+      await tester.pumpAndSettle();
+      sendKeyEventWithCode(19, false, true, false);              // UP_ARROW keyup
+      await tester.pumpAndSettle();
+
+      expect(controller.selection.extentOffset - controller.selection.baseOffset, 12);
+
+      sendKeyEventWithCode(19, true, true, false);               // UP_ARROW keydown
+      await tester.pumpAndSettle();
+      sendKeyEventWithCode(19, false, true, false);              // UP_ARROW keyup
+      await tester.pumpAndSettle();
+
+      expect(controller.selection.extentOffset - controller.selection.baseOffset, 0);
+
+      sendKeyEventWithCode(19, true, true, false);               // UP_ARROW keydown
+      await tester.pumpAndSettle();
+      sendKeyEventWithCode(19, false, true, false);              // UP_ARROW keyup
+      await tester.pumpAndSettle();
+
+      expect(controller.selection.extentOffset - controller.selection.baseOffset, 5);
+    });
+  });
+
```

Comment:
some other tests to have would be:
- if you have two text fields, and both think they are focused, do they both get the events?
- if you have two text fields, and you switch the focus from one to the other and back again, does only the focused one react?
- if you have a text field and you move it from one part of the tree to another using a GlobalKey, does it keep receiving events after?

Reply:
1) That should work right? They should both get events, correct?

Reply:
/ how should I focus both fields? I can't use the tester.tap, because that will only focus one of them...

Reply:
Right, if they both think they're focused they should both get events.
One way to test it would be to have a Column with two MaterialApps inside. Each MaterialApp will starts its own independent focus scope, I believe.

Reply:
Also for 2), what are we trying to test here? What do you mean react? Obviously typing works correctly in this situation.

Reply:
The purpose of tests is to make sure we don't accidentally break things later, not to verify that it works today. I can easily imagine scenarios where we accidentally break things so that switching focus no longer correctly results in the right text field detecting the keystrokes.

Reply:
@Hixie Jeremy and I took a look at this today and we've come to the conclusion that there is no way to set this up in our current test environment. The test binding has a single slot for focus which it maintains, so building two material apps causes the second to entirely steal focus for the first.